### PR TITLE
Module rewritten with new functions and new encryption algorithm

### PR DESCRIPTION
**Pull request by leofreitas**
**Description**
This PR contains a new version with:
- Object oriented rewrite and code cleanup
- Function delete all credentials associated (cascade)
- Convert encryption algorithm mcrypt to openssl
- Added deletion of website logo file

**How Has This Been Tested?**
Local

### CODE REVIEW

**Comment by rossdotparker** [comment]
Hi Leo, this is a massive job, and looks very well done Amazing, thank you! In reviewing the code I can see three issues:

- Some module-specific terms are translated using __() rather than __m()
- There are a lot of white space changes, such as "." to " . ", which I am not sure if PSR compliant. @SKuipers knows more about.
- I did not spot a mechanism for converting existing mycrpt passwords to openssl, so seems like there is no upgrade route. Maybe I missed something?

Let me know what you think, and we'll see what Sandra says. Then we can tweak, merge and test.

File: Credentials/credentials.php [hunk 1]
```diff
@@ -28,142 +32,94 @@
...
+
+    $criteria = $studentGateway->newQueryCriteria()
+            ->searchBy($searchColumns, $search)
+            ->sortBy(['surname', 'preferredName'])
+            ->filterBy('all', $allStudents ?? '')
+            ->fromPOST();
+    $students = $studentGateway->queryStudentBySchoolYear($criteria, $gibbonSchoolYearID);
 
-    if ($result->rowcount() < 1) { echo "<div class='error'>";
+    if (!$students) {
```
Comment (SKuipers) [inline Credentials/credentials.php:80]: DataTables now automatically check for empty result sets, and display a "blank slate" message :) This whole conditional can be removed.

File: Credentials/credentials_student.php [hunk 2]
```diff
@@ -25,134 +32,94 @@
...
+            $studentGateway = $container->get(CredentialsCredentialGateway::class);
+            $searchColumns = $studentGateway->getSearchableColumns();
+
+            $criteria = $studentGateway->newQueryCriteria()
+                            ->searchBy($searchColumns, $search)
+                            ->sortBy(['surname', 'preferredName'])
+                            ->filterBy('all', $allStudents ?? '')
+                            ->fromPOST();
+            $students = $studentGateway->queryStudentBySchoolYear($criteria, $gibbonSchoolYearID, $gibbonPersonID);
+        } catch (Exception $e) {
```
Comment (SKuipers) [inline Credentials/credentials_student.php:56]: Gateway classes now handle exceptions internally, so any try-catch blocks can be removed too :D

File: Credentials/credentials_student.php [hunk 3]
```diff
@@ -25,134 +32,94 @@
...
+                echo "<a href='" . $_SESSION[$guid]['absoluteURL'] . "/index.php?q=/modules/Credentials/credentials.php&search=$search&allStudents=$allStudents'>" . __('Back to Search Results') . '</a>';
                 echo '</div>';
             }
 
             echo "<div class='linkTop'>";
-            echo "<a href='".$_SESSION[$guid]['absoluteURL'].'/index.php?q=/modules/'.$_SESSION[$guid]['module']."/credentials_student_add.php&gibbonPersonID=$gibbonPersonID&search=$search&allStudents=$allStudents'>".__('Add')."<img style='margin-left: 5px' title='".__('Add')."' src='./themes/".$_SESSION[$guid]['gibbonThemeName']."/img/page_new.png'/></a>";
+            echo "<a href='" . $_SESSION[$guid]['absoluteURL'] . '/index.php?q=/modules/' . $_SESSION[$guid]['module'] . "/credentials_student_add.php&gibbonPersonID=$gibbonPersonID&search=$search&allStudents=$allStudents'>" . __('Add') . "<img style='margin-left: 5px' title='" . __('Add') . "' src='./themes/" . $_SESSION[$guid]['gibbonThemeName'] . "/img/page_new.png'/></a>";
```
Comment (SKuipers) [inline Credentials/credentials_student.php:84]: DataTables have an addHeaderAction method, which helps remove these bits of HTML, and takes care of passing values to the action buttons. Eg:

```php
$table->addHeaderAction('add', __('Add'))
      ->addParam('gibbonPersonID', $gibbonPersonID)
      ->addParam('search', $search)
      ->addParam('allStudents', $allStudents)
      ->setURL('/modules/Credentials/credentials_student_add.php')
      ->displayLabel();
```

File: Credentials/credentials_student_addProcess.php [hunk 4]
```diff
@@ -71,21 +55,20 @@
         } else {
             //Write to database
             try {
+                $credentialsCredentialGateway = $container->get(CredentialsCredentialGateway::class);
                 $data = array('gibbonPersonID' => $gibbonPersonID, 'credentialsWebsiteID' => $credentialsWebsiteID, 'username' => $username, 'password' => $passwordFinal, 'notes' => $notes, 'gibbonPersonID' => $gibbonPersonID, 'gibbonPersonIDCreator' => $_SESSION[$guid]['gibbonPersonID'], 'timestampCreator' => date('Y-m-d H:i:s', time()));
-                $sql = 'INSERT INTO credentialsCredential SET credentialsWebsiteID=:credentialsWebsiteID, username=:username, password=:password, notes=:notes, gibbonPersonID=:gibbonPersonID, gibbonPersonIDCreator=:gibbonPersonIDCreator, timestampCreator=:timestampCreator';
-                $result = $connection2->prepare($sql);
-                $result->execute($data);
-            } catch (PDOException $e) {
+                $AI = $credentialsCredentialGateway->insertCredentialsCredential($data);
```
Comment (SKuipers) [inline Credentials/credentials_student_addProcess.php:60]: Gateway classes with the `use TableAware;` trait have some helper methods for insert, update, and delete.

File: Credentials/credentials_student_delete.php [hunk 5]
```diff
@@ -1,94 +1,82 @@
...
-                    $sql = 'SELECT * FROM credentialsCredential WHERE credentialsCredentialID=:credentialsCredentialID';
-                    $result = $connection2->prepare($sql);
-                    $result->execute($data);
-                } catch (PDOException $e) {
-                    echo "<div class='error'>".$e->getMessage().'</div>';
-                }
 
-                if ($result->rowCount() != 1) {
+                $credentialGateway = $container->get(CredentialsCredentialGateway::class);
+                $credential = $credentialGateway->selectCredentialsCredentialById($credentialsCredentialID);
```
Comment (SKuipers) [inline Credentials/credentials_student_delete.php:71]: There's a built-in `getByID()` method that may be handy here.

File: Credentials/credentials_student_deleteProcess.php [hunk 6]
```diff
@@ -1,80 +1,66 @@
...
             header("Location: {$URL}");
         } else {
             //Write to database
             try {
-                $data = array('credentialsCredentialID' => $credentialsCredentialID);
-                $sql = 'DELETE FROM credentialsCredential WHERE credentialsCredentialID=:credentialsCredentialID';
-                $result = $connection2->prepare($sql);
-                $result->execute($data);
-            } catch (PDOException $e) {
+                if (!$credentialGateway->deleteCredentialsCredential($credentialsCredentialID)) {
```
Comment (SKuipers) [inline Credentials/credentials_student_deleteProcess.php:52]: If the `deleteCredentialsCredential` method returns false on failure, the try-catch could perhaps be simplified into an if-else 🤔

File: Credentials/src/CredentialsCredentialGateway.php [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+                ])
+                ->innerJoin('credentialsWebsite', 'credentialsCredential.credentialsWebsiteID=credentialsWebsite.credentialsWebsiteID')
+                ->where("credentialsWebsite.active = 'Y'")
+                ->where("gibbonPersonID=:gibbonPersonID")
+                ->orderBy(['title'])
+                ->bindValue('gibbonPersonID', $gibbonPersonID);     
+        return $this->runQuery($query, $criteria);
+    }
+    
+    public function selectCredentialsCredentialById($credentialsCredentialID) {
```
Comment (SKuipers) [inline Credentials/src/CredentialsCredentialGateway.php:147]: These could be swapped for the built-in `getByID`, `insert`, `update`, `delete' methods.

**Review by SKuipers — COMMENTED** [review]
Wow Leo! You've done an incredible job of figuring out our sparsely-documented classes and putting together this huge refactoring 😃 I've added a couple comments, they're not really errors, just notes to highlight methods that you may not have known where there. Overall this is looking great!

@rossdotparker I think the ` . ` whitespaces aren't necessarily PSR2, but they would fall under an overall code style. Leo, perhaps a quick find-replace could help update them so they're consistent with the core?

**Comment by leofreitas** [comment]
Ross and Sandra,

- Wouldn't it be better to use __m for all terms in the module? Using this method only in some methods does not make it confusing?
- I not found module documentation and support in core to develop a upgrade route. What do you  think of creating something like CHANGEDB to run php scripts in module updates?

**Comment by rossdotparker** [comment]
Hi Leo, for the first point, that is very logical, and I guess that might be how the code works, but @SKuipers would be the best person to confirm.

For the second case, we could look to introduce a system like that, but I think this kind of case is quite rare. Another way to do it would be to add a temporary database field to mark encryption type, and then add some code to the start of CHANGEDB.php that reads all of the rows, and change those marked as mcrypt to openssl. The following version (which could be done in the same release) can then remove the field.

What do you think?

**Comment by rossdotparker** [comment]
@leofreitas long time no see! I hope all is well with you. Just wondering if you have any thoughts on my reply to your second point above. What do you think?

**Comment by leofreitas** [comment]
Hi @rossdotparker, sorry for the delay. I was on vacation, but I came back this week and I'm working on the proposed solution. I think is cool! I will create a new PR.

**Comment by rossdotparker** [comment]
Hi @leofreitas no need to apologise! Thanks for letting us know, and I can't wait to see what you've cooked up.

**Pull request closed without merging**